UDP Ipv6 support

simias opened the pull request:
Far from complete but I could use some feedback

simias commented on src/udp_address.cpp:
```diff
@@ -129,88 +123,55 @@ int zmq::udp_address_t::resolve (const char *name_, bool bind_)
-    return AF_INET;
+    // XXX what do (factor TCP code?)
+    addr_ = address;
+    return 0;
```
What exactly are the requirements for this `to_string` method? The current implementation just returns the string passed to the resolver but the TCP implementation actually bothers re-generating the string from the sockaddr values.

bluca replied:
I have a feeling there was something related to disconnect/unbind - check the git log, there might be some clues

simias commented on src/udp_engine.cpp:
```diff
@@ -111,9 +111,11 @@ void zmq::udp_engine_t::plug (io_thread_t *io_thread_, session_base_t *session_)
         } else {
+            /// XXX fixme ?
             out_address = (sockaddr *) &raw_address;
             out_addrlen = sizeof (sockaddr_in);
```
I don't quite understand how this `raw_address` part fits in, I see that the address is parsed within this very file in `out_event`, where does the string come from exactly? Why doesn't it use the `udp_address` parser?

bluca replied:
not sure about the address, but thw raw_socket option is used to read/write UDP datagrams without any ZMTP headers

simias commented on src/udp_engine.cpp:
```diff
@@ -145,11 +149,29 @@ void zmq::udp_engine_t::plug (io_thread_t *io_thread_, session_base_t *session_)
+                struct ipv6_mreq mreq;
+
+                mreq.ipv6mr_multiaddr = mcast_addr->ipv6.sin6_addr;
+                mreq.ipv6mr_interface = 0; // XXX fixme?
```
So here I'm having a bit of a conundrum: unlike the IPv4 code `struct ipv6_mreq` doesn't let you specify the IP address of the interface to bind but rather the interface number. It's trivial to lookup if we have the interface name using `if_nametoindex` but of course at this point we don't have the interface name anymore and on top of that ZMQ URLs lets you specify an interface by IP: `udp://[abcd::1];[ff00::1]:5555`.

The easy solution would be to only support specifying a source interface by giving the interface name but that's obviously limiting and not very consistent. The hard solution would be to have a method to look up an interface index based on an IP (pretty much the opposite of `zmq::ip_resolver_t::resolve_nic_name`). Of course the tedious part is that this code is very platform-specific.

I think the latter solution is the way to go but I'd like to have some feedback on that.

bluca replied:
If you want to start by only supporting the interface name, and then later on add support for the IP, that's fine given this socket type is a work in progress

bluca commented on unittests/unittest_resolver_common.hpp:
```diff
@@ -0,0 +1,54 @@
+#ifndef __UNITTEST_RESOLVER_COMMON_INCLUDED__
```
a copyright header is needed here

simias commented:
I cleaned up the code, added tests. IPv6 multicast works but only if you bind `*` or an interface name, binding by interface IP returns `ENODEV`. I haven't touched the raw address code so it still only supports IPv4, I suspect that it wouldn't be too hard to modify to use `ip_resolver` instead but it would probably be a good idea to add some tests first.

I want to add multicast tests but I'd need to add a socket option to allow looping the multicast packets on the same host using `setsockopt (IP[V6]_MULTICAST_LOOP)`. I think this ought to be added as a `zmq_setsockopt` parameter but I'll leave that for a future PR. Similarly it might be worth adding support for the `IP_MULTICAST_TTL` option.

simias commented:
I don't understand why some travis builds don't find my header in `unittests`, it works fine on my machine. I tried replacing `#include "unittest_resolver_common.hpp"` with `#include "../unittests/unittest_resolver_common.hpp"` to match the `testutil.hpp` include but it's still broken.

bluca commented:
Doesn't the header have to be added to the executable via CMakeLists.txt?

bluca commented:
The CI is already slow enough, I don't think we should remove the parallel builds - it's easy enough to do in a temporary commit when debugging is needed

bluca commented:
Like this:

https://github.com/zeromq/libzmq/blob/master/tests/CMakeLists.txt#L185

simias commented:
But why does it work when I build locally using cmake?

Also the unnittests CMakeLists.txt doesn't mention any include file, including `testutil.hpp`, should it?

https://github.com/zeromq/libzmq/blob/7aba6821ac56a15706896618ef35d4c9df5032a1/unittests/CMakeLists.txt#L29

simias commented:
Uh, the plot thickens:

```
CMake Error at unittests/CMakeLists.txt:29 (add_executable):
  Cannot find source file:
    unittest_resolver_common.hpp
  Tried extensions .c .C .c++ .cc .cpp .cxx .m .M .mm .h .hh .h++ .hm .hpp
  .hxx .in .txx
```

Does it mean that the buildbot hasn't checked-out that file for some reason?

bluca commented:
gah - it's because the CI tests uses the distributable tarball. You need to add the new header to automake, then it will work

simias commented:
Looks better. If you don't have any objections maybe that's enough changes for this PR and the rest of the work (IPv6 raw sockets, to_string, LOOP support etc...) can be left for later.

simias commented:
Thanks

AntonioGMuriana commented on src/udp_address.cpp:
```diff
@@ -79,24 +77,38 @@ int zmq::udp_address_t::resolve (const char *name_, bool bind_)
+        if (src_name == "*") {
+            bind_interface = 0;
+        } else {
+            bind_interface = if_nametoindex (src_name.c_str ());
```
This breaks again compilation for Universal Windows App.
Reopen issues #2309

AntonioGMuriana replied:
This breaks again compilation for Universal Windows App.
Reopen issue #2309
